radix-transactions/src/manifest/static_resource_movements/typed_invocation.rs

```diff
@@ -1,622 +1,1127 @@
+                    ),*
+                }
 
-                        #[allow(unused_variables, unreachable_code)]
-                        pub fn from_invocation(
-                            function_name: &str,
-                            args: &::radix_common::prelude::ManifestValue
-                        ) -> Result<Self, StaticResourceMovementsError> {
-                            Ok(match function_name {
-                                $(
-                                    $func_name => {
-                                        let encoded = ::radix_common::prelude::manifest_encode(args)
-                                            .map_err(StaticResourceMovementsError::NativeArgumentsEncodeError)?;
-                                        let decoded = ::radix_common::prelude::manifest_decode(&encoded)
-                                            .map_err(StaticResourceMovementsError::NativeArgumentsDecodeError)?;
-                                        Self::$func_ident(decoded)
-                                    }
-                                )*
-                                _ => return Err(StaticResourceMovementsError::UnknownNativeFunction {
-                                    package: $package_address,
-                                    blueprint: stringify!($blueprint_name).to_string(),
-                                    function: function_name.to_string(),
-                                })
+                impl [< $blueprint_ident BlueprintFunction >] {
+                    #![allow(unreachable_patterns, unused_variables)]
+                    pub fn decode_invocation(
+                        function_name: &str,
+                        args: &ManifestValue
+                    ) -> Result<Self, TypedManifestNativeInvocationError> {
+                        match function_name {
+                            $(
+                                $function_name => decode_args(args)
+                                    .map(Self::$function_ident)
+                                    .map_err(|error| {
+                                        TypedManifestNativeInvocationError::FailedToDecodeFunctionInvocation {
+                                            blueprint_id: ::radix_common::prelude::BlueprintId::new(&$package_address, $blueprint_name),
+                                            function_name: function_name.to_owned(),
+                                            args: args.clone(),
+                                            error
+                                        }
+                                    }),
+                            )*
+                            // If we get here then it means that an invalid function was called. We
+                            // have all of the functions on all blueprints we have supported so this
+                            // should be an error.
+                            _ => Err(TypedManifestNativeInvocationError::InvokedFnDoesntExist {
```

Comment:
Can I suggest we spell it out in full as `InvokedFunctionDoesNotExist` or `InvokedFunctionNotFoundOnNativeBlueprint`